Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -298,157 +321,309 @@ void GCMemcardManager::SetSlotFileInteractive(int slot)
+        tr("Successfully exported %1 save files.\nFailed to export %2 save files.")
+            .arg(savefiles.size() - failures)
+            .arg(failures));
+  }
 }
 
-void GCMemcardManager::CopyFiles()
+void GCMemcardManager::ImportFiles(int slot, const std::vector<Memcard::Savefile>& savefiles)
 {
-  auto selection = m_slot_table[m_active_slot]->selectedItems();
-  auto& memcard = m_slot_memcard[m_active_slot];
+  auto& card = m_slot_memcard[slot];
+  if (!card)
+    return;
+
+  const size_t number_of_files = savefiles.size();
+  const size_t number_of_blocks = Memcard::GetBlockCount(savefiles);
+  const size_t free_files = Memcard::DIRLEN - card->GetNumFiles();
+  const size_t free_blocks = card->GetFreeBlocks();
 
-  auto count = selection.count() / m_slot_table[m_active_slot]->columnCount();
+  QStringList error_messages;
 
-  for (int i = 0; i < count; i++)
+  if (number_of_files > free_files)
   {
-    auto sel = selection[i * m_slot_table[m_active_slot]->columnCount()];
-    int file_index = memcard->GetFileIndex(m_slot_table[m_active_slot]->row(sel));
+    error_messages.push_back(tr("There aren't enough free files on the target memory card. At "
+                                "least %1 free files are required.")
+                                 .arg(number_of_files));
+  }
 
-    const auto result = m_slot_memcard[!m_active_slot]->CopyFrom(*memcard, file_index);
+  if (number_of_blocks > free_blocks)
+  {
+    error_messages.push_back(tr("There aren't enough free blocks on the target memory card. At "
+                                "least %1 free blocks are required.")
+                                 .arg(number_of_blocks));
+  }
 
-    if (result != Memcard::GCMemcardImportFileRetVal::SUCCESS)
+  for (const Memcard::Savefile& savefile : savefiles)
+  {
+    if (card->TitlePresent(savefile.dir_entry))
     {
-      ModalMessageBox::warning(this, tr("Copy failed"), tr("Failed to copy file"));
+      const std::string filename = Memcard::GenerateFilename(savefile.dir_entry);
+      error_messages.push_back(tr("The target memory card already contains a file \"%1\".")
+                                   .arg(QString::fromStdString(filename)));
     }
   }
 
-  for (int i = 0; i < SLOT_COUNT; i++)
+  if (!error_messages.empty())
   {
-    if (!m_slot_memcard[i]->Save())
-      PanicAlertT("File write failed");
+    ModalMessageBox::warning(this, tr("Import Failed"), error_messages.join(QLatin1Char{'\n'}));
+    return;
+  }
 
-    UpdateSlotTable(i);
+  for (const Memcard::Savefile& savefile : savefiles)
+  {
+    const auto result = card->ImportFile(savefile);
+
+    // we've already checked everything that could realistically fail here, so this should only
+    // happen if the memory card data is corrupted in some way
+    if (result != Memcard::GCMemcardImportFileRetVal::SUCCESS)
+    {
+      const std::string filename = Memcard::GenerateFilename(savefile.dir_entry);
+      ModalMessageBox::critical(
```

Comment:
These should likely be warnings, given other failures related to importing are also warnings. Also this isn't really crititcal, given it doesn't threaten the stability of the application. It's a problem that can be dealt with.

Ditto for any other potential cases.